src/DockAreaTitleBar.cpp

```diff
@@ -177,27 +199,37 @@ void DockAreaTitleBarPrivate::createButtons()
 
 
     // Close button
-	CloseButton = new tTileBarButton();
-	CloseButton->setObjectName("closeButton");
-	CloseButton->setAutoRaise(true);
-    setTitleBarButtonIcon(CloseButton, QStyle::SP_TitleBarCloseButton, ads::DockAreaCloseIcon);
+    if (testConfigFlag(CDockManager::DockAreaHasCloseButton))
+    {
+        CloseButton = new tTileBarButton();
+        CloseButton->setObjectName("closeButton");
+        CloseButton->setAutoRaise(true);
+        setTitleBarButtonIcon(CloseButton, QStyle::SP_TitleBarCloseButton, ads::DockAreaCloseIcon);
 #ifndef QT_NO_TOOLTIP
-	if (testConfigFlag(CDockManager::DockAreaCloseButtonClosesTab))
-	{
-		CloseButton->setToolTip(QObject::tr("Close Active Tab"));
-	}
-	else
-	{
-		CloseButton->setToolTip(QObject::tr("Close Group"));
-	}
+        if (testConfigFlag(CDockManager::DockAreaCloseButtonClosesTab))
+        {
+            CloseButton->setToolTip(QObject::tr("Close Active Tab"));
+        }
+        else
+        {
+            CloseButton->setToolTip(QObject::tr("Close Group"));
+        }
 #endif
-	CloseButton->setSizePolicy(ButtonSizePolicy);
-	CloseButton->setIconSize(QSize(16, 16));
-	if (testConfigFlag(CDockManager::DockAreaHasCloseButton))
-	{
-		TopLayout->addWidget(CloseButton, 0);
-	}
-	_this->connect(CloseButton, SIGNAL(clicked()), SLOT(onCloseButtonClicked()));
+        CloseButton->setSizePolicy(ButtonSizePolicy);
+        CloseButton->setIconSize(QSize(16, 16));
+        TopLayout->addWidget(CloseButton, 0);
+        _this->connect(CloseButton, SIGNAL(clicked()), SLOT(onCloseButtonClicked()));
+    }
+    // Ideally, we could stop here leaving 'CloseButton' with nullptr value
+    // But there are just too many places in the code where non-null (and non-garbage) value is assumed, so the code crashes without it, so:
+    else
+    {
+        // if CDockManager::DockAreaHasCloseButton configuration is turned off, we'll create a dummy button which is always invisible
+        // (including cases when other code calls it's setVisible() method with 'true')
+        // This hack can be removed when a proper handling of "no CDockManager::DockAreaHasCloseButton" configuration is implemented everywhere in the code
+        CloseButton = new AlwaysHiddenBarButton();
+        CloseButton->hide();
+    }
```

Comment:
I personally think that it should just check that it exists in those places. I think it is only used in a few places inside this file, are you sure you are not finding references to `CDockWidgetTab::CloseButton`?

Reply:
**@rolivav** I totally agree with you that checking data validity before using it is generally the best idea. I have prepared another branch where "AlwaysHiddenBarButton" is not used and that pointer is left with "nullptr". I can offer PR from that branch if you and/or @githubuser0xFFFF will want to see it.

But I still would like to share with you guys the following my considerations:

- Places to be fixed are not in the same file, they are in two other files - "DockAreaWidget.cpp" (1 place)  and "DockContainerWidget.cpp" (3 places)

- The project owner already expressed his (reasonable) preference to receive small isolated PRs. As I wrote in the comment, this hack can be removed once all those places are fixed. I thought that I can do all that once the concept of this PR is accepted, offering all changes in separated small PRs, finally removing the hack.

- While checking pointer against null everywhere is commonly accepted (and probably good) practice it worsens code readability. For example, compare these two:
`DockArea->titleBarButton(TitleBarButtonClose)->setVisible(true);`
with
`auto CloseButton = DockArea->titleBarButton(TitleBarButtonClose);`
`if(CloseButton) CloseButton->setVisible(true);`

- I am planning to introduce also "DockAreaHasCloseButton" flag (for consistency). So:
`DockArea->titleBarButton(TitleBarButtonUndock)->setVisible(true);`
`DockArea->titleBarButton(TitleBarButtonClose)->setVisible(true);`
will become
`auto UndockButton = DockArea->titleBarButton(TitleBarButtonUndock);`
`if(UndockButton) UndockButton->setVisible(true);`
`auto CloseButton = DockArea->titleBarButton(TitleBarButtonClose);`
`if(CloseButton) CloseButton->setVisible(true);`
And  I am planning to introduce also "DockAreaHasMenuButton"...

- As the project will evolve there might be more and more places requiring this check by all project contributors who may simply overlook the possibility for "DockAreaHasCloseButton" config to be false (seems it was overlooked even by the project owner)

- I personally do not like C++ exceptions but one of the reasons behind this invention was making code clean and readable - particularly free of routine checks for nullptr and alike... But why other approaches to avoid those routine checks should be banned? For example, when a C++ class is designed it can guarantee ("by contract") that certain it's getters always return a valid pointer. So users of this class could focus on other activities rather than always validating results of those getters. And if a member of that class can be hidden by that class' object configuration then the code performing the hiding could be fully encapsulated in that class implementation, not exposing additional requirements to class users...

- Other enhancements, like dynamic button show/hide at run-time, having it configured per dock area rather than globally, etc.. will be easier to implement having everything done in the dedicated class.

So I have both approaches implemented in two my branches. I will issue a PR following project owner final decision :)